### PR TITLE
Simd 47 syscall sysvar last restart slot

### PR DESCRIPTION
**Pull request by godmodegalactus**
#### Problem

Implementation of SIMD-47
https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0047-syscall-and-sysvar-for-last-restart-slot.md

#### Summary of Changes

Last restart slot is the last hardfork slot in bank structure.
Creating a sysvar to get last restart slot for the cluster.
Creating a syscall to get last restart slot for the cluster similar to Clock syscall.
Creating the structure `LastRestartSlot` to be used for sysvar.
Implementing the sysvar and updated the sysvar cache.
Adding a feature gate to enable this feature.
Implemented required unit tests.

Fixes #

### CODE REVIEW

File: test.json
Comment (mvines) [inline test.json:1]: Darn!!
```
$ solana balance --url mainnet-beta --keypair test.json 
0 SOL
```

File: sdk/src/feature_set.rs
Comment (mvines) [inline sdk/src/feature_set.rs:669]: Just posting this comment as a reminder: we'll prefer to swap out this keypair for one held by a Solana Labs employee before landing.

File: runtime/src/bank.rs
Comment (mvines) [inline runtime/src/bank.rs:2213]: I have a slight preference to something like this over using a variable named `tmp`
```suggestion
                let hard_forks = self.hard_forks();
                let hard_forks_r = hard_forks.read().unwrap();
                hard_forks_r.iter().last().map(|(slot, _)| *slot).unwrap_or(0)
```

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (mvines) [inline programs/bpf_loader/src/syscalls/mod.rs:12]: This can't be rolled into the `sysvar::{ ... }` block starting on line 7?

File: sdk/program/src/last_restart_slot.rs
Comment (mvines) [inline sdk/program/src/last_restart_slot.rs:5]: `Slot` is already defined in the clock mod:
```suggestion
use crate::clock::Slot;
```

**Review by mvines — COMMENTED** [review]
Overall looks good to me!

@Lichtso or @pgarg66 - can you please recommend a reviewer from runtime to take a look as well

File: sdk/src/feature_set.rs
Comment (godmodegalactus) [reply sdk/src/feature_set.rs:669]: Yes i had already guessed that. I can recommend few keypairs of my own :)

File: program-runtime/src/sysvar_cache.rs
Comment (Lichtso) [inline program-runtime/src/sysvar_cache.rs:181]: nit: inconsistent new line

File: runtime/src/bank/sysvar_cache.rs
Comment (Lichtso) [inline runtime/src/bank/sysvar_cache.rs:11]: I don't think this feature gate here is necessary.

Comment (godmodegalactus) [reply runtime/src/bank/sysvar_cache.rs]: removed the check

File: runtime/src/bank.rs
Comment (godmodegalactus) [inline runtime/src/bank.rs:2220]: @mvines the issue discussed in sdk channel comes from these lines, setting a constant value for last_restart_slot solved the issue.

**Comment by grooviegermanikus** [comment]
I tried to get more insight into this issue using this command "./cargo test --test local_cluster test_hard_fork_with_gap_in_roots" on commit 54d83be7cf on different machines

what I have observed:
* test always passes on my macbook
* test failes some 90%+ on linux laptop and linux server
* yet sometimes test pass on linux !!!

I have collected outputs from the running test for mac, linux-failure, linux-success - see here https://gist.github.com/grooviegermanikus/99ed1e6d57c1b3d0521fb8deee295fb2.
The following log statment indicates if test cluster converges to the newly rooted fork (the count stays at "1" for a while and then increases with slot progression):
`hard fork waiting for 16 new roots.. observed: {AqEWUK8pdsfY2CTrBQLGS8w8ndMeuFcDpCkFwWaicaLL: 1, Bz8byUe5bFKcQZWMdU1NQZuJ2GAN3vZvzkahcynXQi5S: 1}`

`hard fork waiting for 16 new roots.. observed: {AqEWUK8pdsfY2CTrBQLGS8w8ndMeuFcDpCkFwWaicaLL: 1, Bz8byUe5bFKcQZWMdU1NQZuJ2GAN3vZvzkahcynXQi5S: 10}`

loglevels: 
`debug,quinn=warn,hyper=warn,solana_rpc=warn,solana_perf=warn,solana_ledger=warn,solana_metrics::metrics=warn,solana_vote_program::vote_state=info,solana_core::replay_stage=debug,solana_local_cluster=debug,local_cluster=trace`

File: sdk/src/feature_set.rs
Comment (mvines) [reply sdk/src/feature_set.rs:669]:
```suggestion
    solana_sdk::declare_id!("HooKD5NC9QNxk25QuzCssB8ecrEzGt6eXEPBUxWp1LaR");
```
This is a keypair that I hold. I totally didn't grind that prefix!

Comment (mvines) [reply sdk/src/feature_set.rs:669]:
```suggestion
    solana_sdk::declare_id!("HooKD5NC9QNxk25QuzCssB8ecrEzGt6eXEPBUxWp1LaR");
```
This is a keypair that I hold. I totally didn't grind that prefix!

File: test.json
Comment (mvines) [reply test.json:1]: sorry, to be a little more explicit: please remove the test.json file :)

**Comment by mvines** [comment]
@steviez - hey if you have a couple minutes sometime, would you mind taking a look at the local cluster test failure here to see if you maybe spot anything and help these guys out a little

**Comment by steviez** [comment]
Ran out of steam tonight so haven't root-caused it yet, but made some progress:
- I was able to recreate the failure on my linux box (didn't attempt on my mac)
- The logs you provided contained the line `thread panicked while panicking. aborting.`. This is after hitting the timeout assert, which lead me to believe something was panicking earlier (recall that an assert/panic kills a thread, not the process).
- Double checking CI from your PR, I see `thread 'solReplayStage' panicked at 'assertion failed: !last_vote_ancestors.contains(candidate_slot)', core/src/consensus.rs:884:13` in the log
- On my system that was reproducing the failure, I switched the above mentioned line from an assert to `std::process:exit(1)` and confirmed that my local runs were also hitting this

So, as it stand, the test is hitting this assert:
https://github.com/solana-labs/solana/blob/84e28c27f3d14a5e6187e3cb33c564ba869a87e8/core/src/consensus.rs#L881-L884

More so, from looking at logs, I see it is the node that restarted with initial state that is panicking (in the test, this is `validator_b`). `validator_a` is restarting from a manually created snapshot, and not hitting the aforementioned assert.

I didn't spend much time yet comparing logs for mac v. linux, or linux success v linux failure. The fact that the test fails with high frequency on linux is seemingly enough to drill down for root cause. I have some hunches so I can continue looking at this tomorrow, but incase you are interested in pursuing further, two changes I made to speed up debug cycle were
- Reduce that timeout that you bumped up to 300s back down to 100s (I actually plugged in 30s to avoid time waiting)
- Change the assert I linked above to and if statement with `std::process::exit(1)` inside.

**Comment by steviez** [comment]
My latest understanding is that there is some problem outside of the scope of this PR; see details in https://github.com/solana-labs/solana/issues/32117

**Comment by godmodegalactus** [comment]
@mvines @steviez found and fixed the issues.

**Comment by steviez** [comment]
@steviez found and fixed the issues.

I think it is fair to say that I found an issue, but I wouldn't consider it fixed until we fully understand it and have a patch landed in master. The PR doesn't seem related to the test failure; regardless, the test failure is very reproducible on this branch but not the tip of master. I'd feel more comfortable holding off on shipping this PR until we are able to completely absolve it.

**Comment by mvines** [comment]
> The PR doesn't seem related to the test failure; regardless, the test failure is very reproducible on this branch but not the tip of master. I'd feel more comfortable holding off on shipping this PR until we are able to completely absolve it.

sgtm, this PR otherwise looks pretty good and we wouldn't backport this to v1.16, so we have a month or more to get this landed before we'd branch for the v1.17 release line

**Comment by godmodegalactus** [comment]
> @steviez found and fixed the issues.
> 
> I think it is fair to say that I found an issue, but I wouldn't consider it fixed until we fully understand it and have a patch landed in master. The PR doesn't seem related to the test failure; regardless, the test failure is very reproducible on this branch but not the tip of master. I'd feel more comfortable holding off on shipping this PR until we are able to completely absolve it.

> @steviez found and fixed the issues.
> 
> I think it is fair to say that I found an issue, but I wouldn't consider it fixed until we fully understand it and have a patch landed in master. The PR doesn't seem related to the test failure; regardless, the test failure is very reproducible on this branch but not the tip of master. I'd feel more comfortable holding off on shipping this PR until we are able to completely absolve it.

The test no longer fails on this branch, i have tried it multiple times. I think i found out the reason and fixed it.
The fix was that the last restart slot is always less than or equal to the current slot of the bank.

How does it fixes the problem ?

In the test for validator_b we pass through the following code during initialization : 
```

        info!("Processing ledger from genesis");
        let bank_forks = blockstore_processor::process_blockstore_for_bank_0(
            genesis_config,
            blockstore,
            account_paths,
            process_options,
            cache_block_meta_sender,
            entry_notification_sender,
            accounts_update_notifier,
            exit,
        );
```

In the method `process_blockstore_for_bank_0` bank is frozen in the end. And then later we apply hardfork slot in these lines : 
```
if let Some(ref new_hard_forks) = process_options.new_hard_forks {
        let root_bank = bank_forks.read().unwrap().root_bank();
        let hard_forks = root_bank.hard_forks();

        for hard_fork_slot in new_hard_forks.iter() {
            if *hard_fork_slot > root_bank.slot() {
                hard_forks.write().unwrap().register(*hard_fork_slot);
            } else {
                warn!(
                    "Hard fork at {} ignored, --hard-fork option can be removed.",
                    hard_fork_slot
                );
            }
        }
    }
    ```
In the test we have hard fork slot = 40, for bank 0 when we set the sysvar the hard fork was not yet registered so sysvar value is 0 and then hardfork is registered to 40.

For validatorB it goes through different path during concenseus method `wait_for_supermajority` :

```
if let Err(e) = blockstore_processor::process_blockstore_from_root(
                self.blockstore,
                self.bank_forks,
                self.leader_schedule_cache,
                self.process_options,
                self.transaction_status_sender,
                self.cache_block_meta_sender.as_ref(),
                self.entry_notification_sender,
                &self.accounts_background_request_sender,
            ) {
                exit.store(true, Ordering::Relaxed);
                return Err(format!("Failed to load ledger: {e:?}"));
``` 
Here hardfork slot is passed in process_options so bank is initialized with hardfork slot vector with 40 in it. So both sysvar and syscall are set to 40.

Because of these different paths during initialization and concenseus had different values for sysvar because of which they could not find root slot.

So after the fix for bank at slot `0` will always have sysvar set to `0` instead of 40. This is by correct behavior according to specs.

**Comment by steviez** [comment]
>  I think i found out the reason and fixed it.

I believe you have fixed the issue in this PR that was causing the test failure as well, but I would like to elaborate a little. As you pointed out, the two nodes take a different path at this `if` statement in `blockstore_process::load_frozen_forks()`:
https://github.com/solana-labs/solana/blob/4b3045419347ea0757e5a6b5cb009e59a88b0abd/ledger/src/bank_forks_utils.rs#L126-L164

- Node A has previously created a snapshot, so its' `BankForks` will contain a single (root) frozen bank at slot 40
- Node B is loading from genesis. so its' `BankForks` will contain a single (root) frozen bank at slot 0

Later in the same function, any new hard forks will be applied:
https://github.com/solana-labs/solana/blob/4b3045419347ea0757e5a6b5cb009e59a88b0abd/ledger/src/bank_forks_utils.rs#L172-L186

- Node A has no new hard forks in the validator args. However, it is important to remember that Node A is restarting from a snapshot that had the hard fork registered
- Node B will register the new hard fork since the hard fork slot (40) is greater than the root bank (0)

Node B will subsequently replay local ledger state up to slot 40. Node A would have done something similar when it created the snapshot. So, Node A and Node B will have equivalent state at slot 40 when they have each restarted. See below logs from a run with your branch before the fix:
```
// Node A
[2023-06-13T22:04:50.595388979Z INFO  solana_runtime::bank] bank frozen: 40 hash: 7PpacJtvfVZ5dN9928cu2MTH4T8cPAJUDPPK3hrVhtGA ...
// Node B
[2023-06-13T22:05:01.747639923Z INFO  solana_runtime::bank] bank frozen: 40 hash: 7PpacJtvfVZ5dN9928cu2MTH4T8cPAJUDPPK3hrVhtGA ...
```
The bug that you recently identified and fixed was that last restart slot was getting prematurely updated by not ensuring that the hard fork was `<=` the bank's slot. A hard fork in slot 40 should not effect the state in slots `[0, 39]`, but by setting the sysvar early, it actually altered the state for `[1, 39]` (not slot 0 because hard fork applied after bank 0 already frozen). This is illustrated by comparing bank hashes before and after the hard fork:
```
// Node A before restart with hard fork
[2023-06-13T22:04:42.689407246Z INFO  solana_runtime::bank] bank frozen: 39 hash: E7i2EH66i6ddFq34V67VQN2B6WU7yKZn5MA6NzBNMLZz ...

// Node B before restart with hard fork
[2023-06-13T22:04:42.692354218Z INFO  solana_runtime::bank] bank frozen: 39 hash: E7i2EH66i6ddFq34V67VQN2B6WU7yKZn5MA6NzBNMLZz ...

// Node A after hard fork - this bank was frozen while creating restart snapshot
[2023-06-13T22:04:49.195537688Z INFO  solana_runtime::bank] bank frozen: 39 hash: H6TZAZfm85iUACnU3JmVfqELNd42r2qwBqa66iS5D7n3 ...

// Node B after restart with hard fork
[2023-06-13T22:05:01.738018180Z INFO  solana_runtime::bank] bank frozen: 39 hash: H6TZAZfm85iUACnU3JmVfqELNd42r2qwBqa66iS5D7n3 ...
```

While Node A and Node B each had equivalent bank state for slot 40 after restart, they had different local tower state. There are more details in the separate issue that I created (https://github.com/solana-labs/solana/issues/32117), but with the current state of master:
- Node A wipes the local tower due to detecting the hard fork. The tower is late restored by examining the vote that it landed in slot 40
https://github.com/solana-labs/solana/blob/00b5c4012239cb5668692b4129c5fc156a9dd873/core/src/replay_stage.rs#L3001-L3010
- Node B does not detect the hard fork as it relates to the tower, so Node B initializes with the tower it had pre-restart. This tower contains votes for slots beyond slot 40 from before the restart. The votes for pre-restart-slots-over-40 prevent voting for post-restart-slots-over-40, and the state corruption eventually leads to the assertion that I previously linked.

**Comment by godmodegalactus** [comment]
> > I think i found out the reason and fixed it.
> 
> I believe you have fixed the issue in this PR that was causing the test failure as well, but I would like to elaborate a little. As you pointed out, the two nodes take a different path at this `if` statement in `blockstore_process::load_frozen_forks()`:
> 
> https://github.com/solana-labs/solana/blob/4b3045419347ea0757e5a6b5cb009e59a88b0abd/ledger/src/bank_forks_utils.rs#L126-L164
> 
>     * Node A has previously created a snapshot, so its' `BankForks` will contain a single (root) frozen bank at slot 40
> 
>     * Node B is loading from genesis. so its' `BankForks` will contain a single (root) frozen bank at slot 0
> 
> 
> Later in the same function, any new hard forks will be applied:
> 
> https://github.com/solana-labs/solana/blob/4b3045419347ea0757e5a6b5cb009e59a88b0abd/ledger/src/bank_forks_utils.rs#L172-L186
> 
>     * Node A has no new hard forks in the validator args. However, it is important to remember that Node A is restarting from a snapshot that had the hard fork registered
> 
>     * Node B will register the new hard fork since the hard fork slot (40) is greater than the root bank (0)
> 
> 
> Node B will subsequently replay local ledger state up to slot 40. Node A would have done something similar when it created the snapshot. So, Node A and Node B will have equivalent state at slot 40 when they have each restarted. See below logs from a run with your branch before the fix:
> 
> ```
> // Node A
> [2023-06-13T22:04:50.595388979Z INFO  solana_runtime::bank] bank frozen: 40 hash: 7PpacJtvfVZ5dN9928cu2MTH4T8cPAJUDPPK3hrVhtGA ...
> // Node B
> [2023-06-13T22:05:01.747639923Z INFO  solana_runtime::bank] bank frozen: 40 hash: 7PpacJtvfVZ5dN9928cu2MTH4T8cPAJUDPPK3hrVhtGA ...
> ```
> 
> The bug that you recently identified and fixed was that last restart slot was getting prematurely updated by not ensuring that the hard fork was `<=` the bank's slot. A hard fork in slot 40 should not effect the state in slots `[0, 39]`, but by setting the sysvar early, it actually altered the state for `[1, 39]` (not slot 0 because hard fork applied after bank 0 already frozen). This is illustrated by comparing bank hashes before and after the hard fork:
> 
> ```
> // Node A before restart with hard fork
> [2023-06-13T22:04:42.689407246Z INFO  solana_runtime::bank] bank frozen: 39 hash: E7i2EH66i6ddFq34V67VQN2B6WU7yKZn5MA6NzBNMLZz ...
> 
> // Node B before restart with hard fork
> [2023-06-13T22:04:42.692354218Z INFO  solana_runtime::bank] bank frozen: 39 hash: E7i2EH66i6ddFq34V67VQN2B6WU7yKZn5MA6NzBNMLZz ...
> 
> // Node A after hard fork - this bank was frozen while creating restart snapshot
> [2023-06-13T22:04:49.195537688Z INFO  solana_runtime::bank] bank frozen: 39 hash: H6TZAZfm85iUACnU3JmVfqELNd42r2qwBqa66iS5D7n3 ...
> 
> // Node B after restart with hard fork
> [2023-06-13T22:05:01.738018180Z INFO  solana_runtime::bank] bank frozen: 39 hash: H6TZAZfm85iUACnU3JmVfqELNd42r2qwBqa66iS5D7n3 ...
> ```
> 
> While Node A and Node B each had equivalent bank state for slot 40 after restart, they had different local tower state. There are more details in the separate issue that I created (#32117), but with the current state of master:
> 
>     * Node A wipes the local tower due to detecting the hard fork. The tower is late restored by examining the vote that it landed in slot 40
>       https://github.com/solana-labs/solana/blob/00b5c4012239cb5668692b4129c5fc156a9dd873/core/src/replay_stage.rs#L3001-L3010
> 
>     * Node B does not detect the hard fork as it relates to the tower, so Node B initializes with the tower it had pre-restart. This tower contains votes for slots beyond slot 40 from before the restart. The votes for pre-restart-slots-over-40 prevent voting for post-restart-slots-over-40, and the state corruption eventually leads to the assertion that I previously linked.

Yes i agree, there is something to be fixed that is not related to this PR.

**Comment by godmodegalactus** [comment]
Merge?

File: runtime/src/bank.rs
Comment (steviez) [inline runtime/src/bank.rs:2366]: Looks like only half of the comment made it in

File: program-test/tests/sysvar_last_restart_slot.rs
Comment (steviez) [inline program-test/tests/sysvar_last_restart_slot.rs:90]:
```suggestion
    context.warp_to_slot(41).unwrap();
    check_with_program(&mut context, program_id, 41).await;
```
nit: Change to 41 to reinforce that the hard fork at slot `X` is applied in slot `X`

File: program-test/tests/sysvar_last_restart_slot.rs
Comment (steviez) [inline program-test/tests/sysvar_last_restart_slot.rs:92]: Consider the following test case scenario:
```rust
    let mut context = program_test.start_with_context().await;

    check_with_program(&mut context, program_id, 0).await;
    context.warp_to_slot(39).unwrap();
    context.register_hard_fork(20 as Slot);
    context.warp_to_slot(40).unwrap();

    check_with_program(&mut context, program_id, 20).await;
```
- For the validator codepath, `bank_fork_utils::load_bank_forks()` ignores hard forks that are <= the snapshot slot (or 0 if we were loading from genesis).
- The `HardForks` struct itself does not impose such a restriction, and I believe 20 would be registered just fine. This would then lead to 20 being set in last_restart_slot variable.

I'm not sure this difference is a concrete problem, but at the very least, it could definitely be confusing. One option to enforce consistent behavior would be passing the bank (or at least its' slot) into `HardForks::register()`. Any thoughts on this @mvines ?

File: runtime/src/bank.rs
Comment (godmodegalactus) [reply runtime/src/bank.rs:2366]: Ahh yes silly me the second line did not `// applying a hard fork that is set to occur in the future.` make sense to me just reading it standalone, it makes more sense combining with the first line.

Comment (godmodegalactus) [reply runtime/src/bank.rs:2366]: I will fix it ASAP.

File: program-test/tests/sysvar_last_restart_slot.rs
Comment (mvines) [reply program-test/tests/sysvar_last_restart_slot.rs:92]: `context.register_hard_fork(20 as Slot);` to me is nonsensical in this context. Actually panicking on an attempt to register a hard fork older than the working bank + 1 seems fine for example.

Comment (godmodegalactus) [reply program-test/tests/sysvar_last_restart_slot.rs:92]: Currently there is no panic when we attempt to register older hardfork. So we can always set an older hardfork.
Adding a panic seems to be out of context for this PR and may add some more issues, but happy to create a new issue for that.

https://github.com/solana-labs/solana/blob/47ff3cecc9cad527b208836f39a7321bf0511b5e/sdk/src/hard_forks.rs#LL18C8-L18C8

Comment (mvines) [reply program-test/tests/sysvar_last_restart_slot.rs:92]: thanks, i think it's fine to not add an explicit panic in that case

File: runtime/src/bank/sysvar_cache.rs
Comment (mvines) [inline runtime/src/bank/sysvar_cache.rs:10]: Why was sysvar_cache.rs changed? The new code seems equivalent, but more wordy

**Review by mvines — APPROVED** [review]
looks good to me, I've created a feature gate issue as I hold the activation keypair: https://github.com/solana-labs/solana/issues/32177

@steviez - any remaining concerns about this landing in master?

**Comment by steviez** [comment]
> any remaining concerns about this landing in master?

@mvines - No remaining concerns for me, I can take a stab at addressing the item I raised https://github.com/solana-labs/solana/pull/31957#discussion_r1231489773 once this PR lands

File: runtime/src/bank/sysvar_cache.rs
Comment (godmodegalactus) [reply runtime/src/bank/sysvar_cache.rs:10]: There was a check here which was removed following a comment. I will revert this file then.

**Review by mvines — APPROVED** [review]
thanks for working on this!

**Pull request merged**
